servlet/src/main/java/io/undertow/servlet/spec/ServletOutputStreamImpl.java

```diff
@@ -868,27 +874,30 @@ public void handleEvent(final StreamSinkChannel aChannel) {
                 }
 
                 state |= FLAG_READY;
-                try {
-                    state |= FLAG_IN_CALLBACK;
-
-                    servletRequestContext.getCurrentServletContext().invokeOnWritePossible(servletRequestContext.getExchange(), listener);
-
-                    if (isReady()) {
-                        //if the stream is still ready then we do not resume writes
-                        //this is per spec, we only call the listener once for each time
-                        //isReady returns true
-                        if (channel != null) {
-                            channel.suspendWrites();
-                        }
-                    } else {
-                        if (channel != null) {
-                            channel.resumeWrites();
+                if (anyAreSet(state, FLAG_RECEIVED_NOT_READY)) {
+                    state &= ~FLAG_RECEIVED_NOT_READY;
+                    try {
+                        state |= FLAG_IN_CALLBACK;
+
+                        servletRequestContext.getCurrentServletContext().invokeOnWritePossible(servletRequestContext.getExchange(), listener);
+
+                        if (anyAreSet(state, FLAG_READY)) {
+                            //if the stream is still ready then we do not resume writes
+                            //this is per spec, we only call the listener once for each time
+                            //isReady returns true
+                            if (channel != null) {
+                                channel.suspendWrites();
+                            }
```

Comment:
I'm not sure if this branch of the conditional is still necessary. Perhaps this would work?
```java
if (!anyAreSet(state, FLAG_READY) && channel != null) {
    channel.resumeWrites();
}
```